File: manual/code/test_suites/cpp/FunctionBuildTest/src/function_spacy_pos.cpp
#include <iostream>
#include <string>
#include <vector>
#include <stdexcept>
#include <stanfordnlp/StanfordCoreNLP.h>
#include <stanfordnlp/Annotation.h>
#include <stanfordnlp/CoreAnnotations.h>
#include <stanfordnlp/CoreLabel.h>
#include <stanfordnlp/CoreMap.h>

class SpacyPosProcessor {
public:
    std::string process(const std::string& text) {
        if (text.empty()) {
            throw std::invalid_argument("The input must be of string type.");
        }

        try {
            // 配置StanfordCoreNLP，添加ssplit注释器
            std::map<std::string, std::string> props = {{"annotators", "tokenize, ssplit, pos"}};
            StanfordCoreNLP pipeline(props);

            // 创建注释对象
            Annotation document(text);
            pipeline.annotate(document);

            // 获取句子列表
            std::vector<CoreMap> sentences = document.get<CoreAnnotations::SentencesAnnotation>();

            // 构建POS标注结果
            std::string posTags;
            for (const auto& sentence : sentences) {
                for (const auto& token : sentence.get<CoreAnnotations::TokensAnnotation>()) {
                    std::string word = token.word();
                    std::string pos = token.get<CoreAnnotations::PartOfSpeechAnnotation>();
                    posTags += word + " (" + pos + "), ";
                }
            }

            // 移除最后的逗号和空格
            if (!posTags.empty()) {
                posTags = posTags.substr(0, posTags.length() - 2);
            }

            return posTags;

        } catch (const std::exception& e) {
            std::cerr << "OtherError: " << e.what() << std::endl;
            throw;
        }
    }
};

class FunctionSpacyPos {
public:
    std::string test_pos(const std::string& text) {
        SpacyPosProcessor processor;
        return processor.process(text);
    }
};

File: manual/data/java/src/function_spacy_pos.java
package com.example;

import edu.stanford.nlp.ling.CoreAnnotations;
import edu.stanford.nlp.ling.CoreLabel;
import edu.stanford.nlp.pipeline.Annotation;
import edu.stanford.nlp.pipeline.StanfordCoreNLP;
import edu.stanford.nlp.util.CoreMap;

import java.util.Properties;
import java.util.List;
import java.util.StringJoiner;

public class SpacyPosProcessor {
    public String test_pos(String text) {
        class SpacyPosProcessor {
            private StanfordCoreNLP pipeline;

            public SpacyPosProcessor() {
                // 初始化pipeline
                Properties props = new Properties();
                props.setProperty("annotators", "tokenize, ssplit, pos");
                pipeline = new StanfordCoreNLP(props);
            }

            public String processText(String inputText) {
                if (inputText == null) {
                    throw new IllegalArgumentException("The input must be of string type.");
                }

                try {
                    // 创建注释对象
                    Annotation document = new Annotation(inputText);
                    pipeline.annotate(document);

                    // 获取句子
                    List<CoreMap> sentences = document.get(CoreAnnotations.SentencesAnnotation.class);
                    StringJoiner posTagsJoiner = new StringJoiner(", ");

                    // 处理每个句子中的token
                    for (CoreMap sentence : sentences) {
                        for (CoreLabel token : sentence.get(CoreAnnotations.TokensAnnotation.class)) {
                            String word = token.word();
                            String pos = token.get(CoreAnnotations.PartOfSpeechAnnotation.class);
                            String mappedPos = mapPosTag(pos);
                            if (!mappedPos.isEmpty()) {
                                posTagsJoiner.add(word + " (" + mappedPos + ")");
                            }
                        }
                    }

                    return posTagsJoiner.toString();
                } catch (Exception e) {
                    System.out.println("OtherError: " + e.getMessage());
                    throw e;
                }
            }

            private String mapPosTag(String stanfordTag) {
                // 扩展映射以更好地匹配spaCy的标签
                return switch (stanfordTag) {
                    // Nouns
                    case "NN", "NNS", "NNP", "NNPS" -> "NOUN";
                    // Verbs
                    case "VB", "VBD", "VBG", "VBN", "VBP", "VBZ" -> "VERB";
                    // Adjectives
                    case "JJ", "JJR", "JJS" -> "ADJ";
                    // Adverbs
                    case "RB", "RBR", "RBS" -> "ADV";
                    // Punctuation
                    case ".", ",", ":", ";", "!", "?" -> "PUNCT";
                    // Articles and Determiners
                    case "DT" -> "DET";
                    // Pronouns
                    case "PRP", "PRP$", "WP", "WP$" -> "PRON";
                    // Prepositions
                    case "IN" -> "ADP";
                    // Conjunctions
                    case "CC" -> "CONJ";
                    // Modal verbs
                    case "MD" -> "VERB";
                    // Default case
                    default -> stanfordTag;
                };
            }
        }

        SpacyPosProcessor processor = new SpacyPosProcessor();
        return processor.processText(text);
    }
}